drivers/tsl4531x/tsl4531x.c

```diff
@@ -136,62 +136,93 @@ int tsl4531x_start_sample(tsl4531x_t *dev)
 {
     assert(dev);
 
+    int r;
+
     /* Don't change the mode to one-shot if the device is in high power mode. */
     if (dev->low_power_mode) {
-        _start_sample(dev);
+
+        if ((r = i2c_acquire(dev->i2c_dev)) < 0) {
+            DEBUG("[Error] Cannot acquire device. I2C error: %d\n", r);
+            return -ENODEV;
+        }
+
+        if ((r = i2c_write_reg(dev->i2c_dev,
+                               dev->i2c_addr,
+                               TSL4531X_COMMAND(TSL4531X_CONTROL_REG),
+                               TSL4531X_CONTROL(TSL4531X_MODE_SINGLE_ADC_CYCLE),
+                               0)) < 0) {
+            DEBUG("[Error] Cannot write power mode. I2C error: %d\n", r);
+            i2c_release(dev->i2c_dev);
+            return -ENXIO;
+        }
+
+        i2c_release(dev->i2c_dev);
+
         dev->sample_start_time = xtimer_now_usec();
     }
 
     return 0;
 }
 
-uint32_t tsl4531x_time_until_sample_ready(const tsl4531x_t *dev)
+uint32_t tsl4531x_time_until_sample_ready(tsl4531x_t *dev)
 {
     assert(dev);
 
-    uint32_t elapsed = (xtimer_now_usec() - dev->sample_start_time);
+    uint32_t t = TSL4531X_GET_INTEGRATION_TIME_USEC(dev->integration_time, TSL4531X_PSAVESKIP_ON) -
+                 (xtimer_now_usec() - dev->sample_start_time);
 
-    return (TSL4531X_GET_INTEGRATION_TIME_USEC(dev->integration_time, TSL4531X_PSAVESKIP_ON) > elapsed ?
-            TSL4531X_GET_INTEGRATION_TIME_USEC(dev->integration_time, TSL4531X_PSAVESKIP_ON) - elapsed :
-            0);
+    /* Clamp t at zero */
+    t = (t <= TSL4531X_GET_INTEGRATION_TIME_USEC(dev->integration_time, TSL4531X_PSAVESKIP_ON) ?
```

Comment:
if 
* `t = TSL4531X_GET_INTEGRATION_TIME_USEC(...) - (xtimer_now_usec() - dev->sample_start_time);`
* `(xtimer_now_usec() - dev->sample_start_time) > 0` (which is always true because "now" is always later than the time we started the sample)
then
* `t <= TSL4531X_GET_INTEGRATION_TIME_USEC(...)` is always true.

Reply:
However, if 
`(xtimer_now_usec() - dev->sample_start_time)`
is greater than 
`TSL4531X_GET_INTEGRATION_TIME_USEC(...)`
then t becomes negative, which wraps around to a large positive number because t is unsigned.
so,
`t > TSL4531X_GET_INTEGRATION_TIME_USEC(...)`. at that point.